controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -522,6 +528,24 @@ func (r *BareMetalHostReconciler) actionUnmanaged(prov provisioner.Provisioner,
 	return actionContinue{unmanagedRetryDelay}
 }
 
+func (r *BareMetalHostReconciler) actionDetached(prov provisioner.Provisioner, info *reconcileInfo) actionResult {
+	provResult, err := prov.Detach()
+	if err != nil {
+		return actionError{errors.Wrap(err, "failed to detach")}
+	}
+	if provResult.ErrorMessage != "" {
+		return recordActionFailure(info, metal3v1alpha1.DetachError, provResult.ErrorMessage)
+	}
+	if provResult.Dirty {
+		return actionContinue{provResult.RequeueAfter}
+	}
+	if clearErrorSetStatus(info.host, metal3v1alpha1.OperationalStatusDetached) {
+		info.log.Info("host is detached, removed from provisioner")
```

Comment:
I guess this could be a good point to reset the ErrorCount to 0

Reply:
Thanks @andfasano - the only place I see `info.host.Status.ErrorCount = 0` is in `manageHostPower` - do we expect that to reset the counter for all non-detached cases?  I incorrectly assumed it was handled by the existing `clearError` code

Reply:
Resolved in the latest version, thanks!

Reply:
Originally it was defined there, but then the `ErrorCount` resetting has been mostly moved into the `host_state_machine` (usually on state transitions as [here](https://github.com/metal3-io/baremetal-operator/blob/b7cc13f55e3e8db9fa0318b71b4c64b6d4d14fe0/controllers/metal3.io/host_state_machine.go#L300) or [here](https://github.com/metal3-io/baremetal-operator/blob/b7cc13f55e3e8db9fa0318b71b4c64b6d4d14fe0/controllers/metal3.io/host_state_machine.go#L308) for example).
In this specific case we are exercising an action - `Detach` - that could fail and increase the error count, so as long as the resource remains detached (with the annotation set and state `Provisioned` or `ExternallyProvisioned`),_ a previously failed - but now recovered - deletion can be cleared here when the `operationalStatus` becomes effectively `OperationalStatusDetached`.

I just realized that there could be another scenario to be handled regarding the error management IIUC the annotation handling (@hardys can you please confirm if it sound reasonable?): when the `actionDetached` is triggered, the `Detach` fails (so the `ErrorCount` is increased) but then the annotation is removed (by a human or any other external operator) before recovering. The resource is still marked to be in error (`ErrorCount` increased, `ErrorType` set to `DetachError` and `ErrorMessage` no empty), but the flow at this point will go back as usual in the `actionManageSteadyState`.

Reply:
@andfasano right, so in that case we'll call `manage[Externally]Provisioned->actionManageSteadyState->manageHostPower` which should reset the `ErrorCount` right?

I think reproducing such a scenario will be difficult, but I can try to add a test if you think it's needed?

Reply:
Yes, but the main issue I see is that potentially other problems could pop out before reaching that point (ie maybe the `Adopt` fails), and while the `ErrorType` and `ErrorMessage` will be overwritten the `ErrorCount` will get increased from its current non-zero value (non-zero due another unrelated error).

I'd say that for the sake of clarity it could be cleaner to detect the "interruption" of a in-progress detach operation (annotation not set but `ErrorType` set to `DetachError`) [here](https://github.com/metal3-io/baremetal-operator/pull/827/files#diff-eeed3703d04e4c23a7d7af8cd0b7931b6b7990f23d826c49bdbc31c5f0a50291R251) and perform the cleanup, since it looks like the detaching has been abandoned.

Having a test to capture it would be really great.

Reply:
Ok I added a commit which adds a test for the two error paths, e.g:

  * Detach() fails, then succeeds
  * Detach() fails, then the annotation is removed

In the latter case I reset the ErrorCount, although I guess it's pretty likely whatever caused the Detach() to fail will cause some subsequent error.